Mathlib/Topology/Order/Hom/Basic.lean

```diff
@@ -0,0 +1,199 @@
+/-
+Copyright (c) 2022 Yaël Dillies. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies
+
+! This file was ported from Lean 3 source module topology.order.hom.basic
+! leanprover-community/mathlib commit 0a0ec35061ed9960bf0e7ffb0335f44447b58977
+! Please do not edit these lines, except to modify the commit id
+! if you have ported upstream changes.
+-/
+import Mathlib.Order.Hom.Basic
+import Mathlib.Topology.ContinuousFunction.Basic
+
+/-!
+# Continuous order homomorphisms
+
+This file defines continuous order homomorphisms, that is maps which are both continuous and
+monotone. They are also called Priestley homomorphisms because they are the morphisms of the
+category of Priestley spaces.
+
+We use the `FunLike` design, so each type of morphisms has a companion typeclass which is meant to
+be satisfied by itself and all stricter types.
+
+## Types of morphisms
+
+* `ContinuousOrderHom`: Continuous monotone functions, aka Priestley homomorphisms.
+
+## Typeclasses
+
+* `ContinuousOrderHomClass`
+-/
+
+
+open Function
+
+variable {F α β γ δ : Type _}
+
+/-- The type of continuous monotone maps from `α` to `β`, aka Priestley homomorphisms. -/
+structure ContinuousOrderHom (α β : Type _) [Preorder α] [Preorder β] [TopologicalSpace α]
+  [TopologicalSpace β] extends OrderHom α β where
+  continuous_toFun : Continuous toFun
+#align continuous_order_hom ContinuousOrderHom
+
+-- mathport name: «expr →Co »
+infixr:25 " →Co " => ContinuousOrderHom
+
+section
+
+-- porting note: extending `ContinuousMapClass` instead of `OrderHomClass`
+/-- `ContinuousOrderHomClass F α β` states that `F` is a type of continuous monotone maps.
+
+You should extend this class when you extend `ContinuousOrderHom`. -/
+class ContinuousOrderHomClass (F : Type _) (α β : outParam <| Type _) [Preorder α] [Preorder β]
+    [TopologicalSpace α] [TopologicalSpace β] extends
+    ContinuousMapClass F α β where
+  map_monotone (f : F) : Monotone f
+#align continuous_order_hom_class ContinuousOrderHomClass
+
+end
+
+-- See note [lower instance priority]
+instance (priority := 100) ContinuousOrderHomClass.toOrderHomClass [Preorder α] [Preorder β]
+    [TopologicalSpace α] [TopologicalSpace β] [ContinuousOrderHomClass F α β] :
```

Comment:
Wouldn't this become a dangerous instance because the `Preorder` and `TopologicalSpace` parameters depend on an `outParam` in `OrderHomClass` but are synthesized before `ContinuousOrderHomClass` can fill them in? It would explain the timeouts, where Lean would proceed to try all possible instances for `Preorder ?α` on a free metavariable `?α`.

So I think the following is safer:
```suggestion
instance (priority := 100) ContinuousOrderHomClass.toOrderHomClass {_ : Preorder α} {_ : Preorder β}
    {_ : TopologicalSpace α} {_ : TopologicalSpace β} [ContinuousOrderHomClass F α β] :
```
This should work fine since the `ContinuousOrderHomClass` instance can fill in the implicits.

Reply:
You are completely right, Anne.  This should have been caught by the linter. https://github.com/leanprover/std4/pull/102